File: chess_package/bot/starter_code/c/bot.c
/* bot.c
 * Code Clash Chess Challenge — C Starter Bot
 *
 * Usage:
 *   gcc -std=c11 bot.c cJSON.c -o bot
 *   ./bot /path/to/state.json
 *
 * Output:
 *   Writes `move.json` in the current directory.
 */

#include <stdio.h>
#include <stdlib.h>
#include "cJSON.h"
#include "constants.h"

// Load the full JSON state from disk
cJSON *load_state(const char *path) {
    FILE *f = fopen(path, "rb");
    if (!f) {
        fprintf(stderr, "Error: cannot open %s\n", path);
        exit(1);
    }
    fseek(f, 0, SEEK_END);
    long len = ftell(f);
    fseek(f, 0, SEEK_SET);
    char *data = malloc(len + 1);
    fread(data, 1, len, f);
    data[len] = '\0';
    fclose(f);

    cJSON *state = cJSON_Parse(data);
    free(data);
    if (!state) {
        fprintf(stderr, "Error: invalid JSON in %s\n", path);
        exit(2);
    }
    return state;
}

// Write the bot's chosen move into move.json
void write_move(cJSON *move) {
    FILE *f = fopen("move.json", "w");
    if (!f) {
        fprintf(stderr, "Error: cannot write move.json\n");
        exit(3);
    }
    char *out = cJSON_Print(move);
    fprintf(f, "%s\n", out);
    fclose(f);
    cJSON_free(out);
}

// Called when phase == "setup"
cJSON *setup_phase(cJSON *state) {
    /*
     * TODO: implement your setup logic.
     * state->["setupStep"], ["playerColor"], ["board"], ["blockedTiles"], etc.
     *
     * return a JSON object like:
     * {
     *   "move": { "from": [x,y], "to": [r,c] },
     *   "ability": { "name": NULL, "target": NULL }
     * }
     */
    cJSON *move = cJSON_CreateObject();
    cJSON *mv = cJSON_CreateObject();
    cJSON_AddItemToObject(move, "move", mv);
    cJSON_AddItemToObject(mv, "from", cJSON_CreateNull());
    cJSON_AddItemToObject(mv, "to", cJSON_CreateIntArray((int[]){0,0}, 2));

    cJSON *ab = cJSON_CreateObject();
    cJSON_AddItemToObject(move, "ability", ab);
    cJSON_AddItemToObject(ab, "name", cJSON_CreateNull());
    cJSON_AddItemToObject(ab, "target", cJSON_CreateNull());

    return move;
}

// Called when phase == "play"
cJSON *play_phase(cJSON *state) {
    /*
     * TODO: implement your play logic.
     * state->["phase"], ["board"], ["abilitiesRemaining"], etc.
     *
     * return a JSON object like:
     * {
     *   "move": { "from": [r1,c1], "to": [r2,c2] },
     *   "ability": { "name": NULL, "target": NULL }
     * }
     */
    cJSON *move = cJSON_CreateObject();
    cJSON *mv = cJSON_CreateObject();
    cJSON_AddItemToObject(move, "move", mv);
    cJSON_AddItemToObject(mv, "from", cJSON_CreateIntArray((int[]){0,0}, 2));
    cJSON_AddItemToObject(mv, "to", cJSON_CreateIntArray((int[]){0,1}, 2));

    cJSON *ab = cJSON_CreateObject();
    cJSON_AddItemToObject(move, "ability", ab);
    cJSON_AddItemToObject(ab, "name", cJSON_CreateNull());
    cJSON_AddItemToObject(ab, "target", cJSON_CreateNull());

    return move;
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s /path/to/state.json\n", argv[0]);
        return 1;
    }

    cJSON *state = load_state(argv[1]);
    const cJSON *phase = cJSON_GetObjectItemCaseSensitive(state, "phase");
    if (!cJSON_IsString(phase)) {
        fprintf(stderr, "Error: missing phase in state.json\n");
        return 2;
    }

    cJSON *move = NULL;
    if (strcmp(phase->valuestring, "setup") == 0) {
        move = setup_phase(state);
    } else {
        move = play_phase(state);
    }

    write_move(move);

    cJSON_Delete(move);
    cJSON_Delete(state);
    return 0;
}

File: chess_package/bot/starter_code/python/bot.py
#!/usr/bin/env python3
"""
Code Clash Chess Challenge — Python Starter Bot

Welcome to the Code Clash Chess Challenge! This is the official Python starter bot for the
custom 5×5 chess variant used in the competition.

You're free to change **any** part of this code. It's just here to help you get started.

-------------------------------------------
How to package your bot as a single file:
-------------------------------------------
1. Install PyInstaller (if you don’t already have it):
   pip install pyinstaller

2. Package your bot into a one-file executable:
   pyinstaller --onefile bot.py

3. The resulting executable will be located in the `dist/` folder
   (e.g., `dist/bot.exe` on Windows)

For rules and move formats, see the `design_doc.md` file provided.
"""

import json
import sys
from typing import Any, Dict

BOARD_SIZE = 5


def on_board(pos) -> bool:
    """Check if a position is within the board boundaries."""
    r, c = pos
    return 0 <= r < BOARD_SIZE and 0 <= c < BOARD_SIZE


def load_state(path: str) -> Dict[str, Any]:
    """Load and return the game state from the given JSON file."""
    with open(path, 'r', encoding='utf-8') as f:
        return json.load(f)


def write_move(move_data: Dict[str, Any]) -> None:
    """Write the chosen move (and ability) to move.json."""
    with open('move.json', 'w', encoding='utf-8') as f:
        json.dump(move_data, f)


def setup_phase(state: Dict[str, Any]) -> Dict[str, Any]:
    """
    Implement your setup phase strategy here.

    Step 1: Place King
    Step 2: Block opponent tile
    Step 3: Place 2 Rooks
    Step 4: Place 2 Bishops and 3 Pawns

    Return a dict: {'move': {'from': [...], 'to': [row, col]}}
    """
    # TODO: Implement your setup logic
    raise NotImplementedError("setup_phase() not implemented yet")


def play_phase(state: Dict[str, Any]) -> Dict[str, Any]:
    """
    Implement your play phase strategy here.

    Choose a legal move and optionally activate one ability.

    Return a dict like:
    {
        'move': {'from': [row, col], 'to': [row, col]},
        'ability': {'name': 'fog' | 'pawnReset' | 'shield' | None, 'target': [row, col] | None}
    }
    """
    # TODO: Implement your play logic
    raise NotImplementedError("play_phase() not implemented yet")


def main():
    """
    Entry point for the bot.

    Reads state from file, calls the appropriate phase function, and writes move.json.
    """
    if len(sys.argv) != 2:
        print("Usage: python bot.py /path/to/state.json")
        print("Usage (compiled): ./bot /path/to/state.json")
        sys.exit(1)

    state_path = sys.argv[1]

    try:
        state = load_state(state_path)
    except Exception:
        sys.exit(1)

    try:
        phase = state.get('phase')
        if phase == 'setup':
            move_data = setup_phase(state)
        else:
            move_data = play_phase(state)

        write_move(move_data)
        sys.exit(0)
    except Exception:
        sys.exit(3)


if __name__ == '__main__':
    main()